python/mlc_chat/support/preshard.py

```diff
@@ -133,3 +137,162 @@ def apply_preshard(
+                    a_scales =  _split_array(scales_dict[a_scale], tensor_parallel_shards)
+                    w_scales =  _duplicate_array(scales_dict[w_scale], tensor_parallel_shards)
+                    a_zps =     _split_array(scales_dict[a_zp], tensor_parallel_shards)
+                    w_zps =     _duplicate_array(scales_dict[w_zp], tensor_parallel_shards)
+                for shard_idx in range(tensor_parallel_shards):
+                    out[_sharded_param_name(a_factor, shard_idx)] = a_factors[shard_idx]
+                    out[_sharded_param_name(w_factor, shard_idx)] = w_factors[shard_idx]
+                    if not args.quantization.name in smooth_0_quants:
+                        out[_sharded_param_name(w_scale, shard_idx)] = w_scales[shard_idx]
+                        out[_sharded_param_name(w_zp, shard_idx)] = w_zps[shard_idx]
+            else:
+                out[a_factor] = smoothing_factors_dict[a_factor]
+                out[w_factor] = smoothing_factors_dict[w_factor]
+                if not args.quantization.name in smooth_0_quants:
+                    out[w_scale]  = scales_dict[w_scale]
+                    out[w_zp]  = scales_dict[w_zp]
+    return out
+
+def load_file(path):
+    import json
+    with open(path, 'r') as f:
+        loaded_dict = json.load(f)
+    return loaded_dict
+
+def _create_smoothquant_func(
+    bb: relax.BlockBuilder, param: nn.Parameter, param_name: str, idx: int, tensor_parallel_shards: int, **smq_params
+):
+    def _create_func(
+        func_name: str,
+        bb: relax.BlockBuilder,
+        param: nn.Parameter,
+        smoothing_factor: Union[np.ndarray, nd.NDArray],
+        scale: Union[np.ndarray, nd.NDArray],
+        zp: Union[np.ndarray, nd.NDArray],
+        dtype: str
+    ):
+        weight_var = relax.Var("weight", relax.TensorStructInfo(param.shape, param.dtype))
+        with bb.function(name=func_name, params=[weight_var]):
+            with bb.dataflow():
+                if smoothing_factor is not None:
+                    weight_var = bb.emit(relax.op.multiply(weight_var, relax.const(smoothing_factor)))
+                if scale is not None and zp is not None:
+                    weight_var = bb.emit(relax.op.quantize(weight_var, relax.const(scale), relax.const(zp), axis=-2, out_dtype=dtype))
+                gv = bb.emit_output(weight_var)
+            bb.emit_func_output(gv)
+
+    func_names = []
+    shard_strategy = param.attrs.get("shard_strategy", None)
+    factor_param = smq_params.get("smoothing_factor")
+    scale_param = smq_params.get("scale")
+    zp_param = smq_params.get("zp")
+    if tensor_parallel_shards == 1 or shard_strategy is None:
+        func_name = f"convert_param_{idx}"
+        func_names.append((param_name, func_name))
+        _create_func(func_name, bb, param, factor_param, scale_param, zp_param, smq_params.get("quant_config").weight_dtype)
+    else:
+        if shard_strategy.dim == 0:
```

Comment:
Is the goal here to apply shading to scales / zps? can we attach shard strategy to these parameters so that these functions can be generated automatically

Reply:
we may refactor after we bring in https://github.com/mlc-ai/mlc-llm/pull/2039 since it changes the order of applying presharding and may require a different shard strategy

Reply:
Yes, looks like shard strategy should be changed if this PR will be landed.

Reply:
I've left this code as it is for now. If [mlc-ai#2039](https://github.com/mlc-ai/mlc-llm/pull/2039) will be landed, this sharding will not be required.